### ISSUE TITLE
Private init/destroy method may be invoked twice

### ISSUE THREAD

**Issue by levitin** [spring-projects/spring-framework#28083]
Imagine, you have a bean with an `init` method, annotated with `@PostConstruct` and a configuration class, referencing the same `init` method on its own.

```java
public class MyBean {

    @PostConstruct
    public void init() {
        ...
    }

}
```

```java
@Configuration
public class MyConfiguration {

    @Bean(initMethod = "init")
    public MyBean myBean() {
        return new MyBean();
    }
}
```

The `init` method will be invoked only once (what is pretty well). 

This is true for all method modifiers except `private`.
As soon as you change the method modifier to `private` the  `init` method is invoked twice. 

```diff
public class MyBean {

    @PostConstruct
-    public void init() {
+    private void init() {
        ...
    }
}
```

**Comment by MrocCyen** [comment]
Hi, I test it and maked a same result, I found the problem.

1、when it is private
![image](https://user-images.githubusercontent.com/22514893/154883386-b9432c11-bf1e-4227-8d38-ab1398daf127.png)
![image](https://user-images.githubusercontent.com/22514893/154883697-a9a5b28c-dccb-49e3-ab42-d9d180bc8953.png)
![image](https://user-images.githubusercontent.com/22514893/154883826-f80d4dea-0079-4b16-84df-40eacbbdc642.png)

externallyManagedInitMethods value is 'org.springframework.issues.issue_28083.MyBean.init', but initMethod value is 'init',  invokeCustomInitMethod method will be invoked

2、when it is not private
![image](https://user-images.githubusercontent.com/22514893/154884230-676ab3d7-e464-4b35-b45f-d9743655390f.png)
![image](https://user-images.githubusercontent.com/22514893/154884379-06f1d9b4-cdbb-47ed-975b-41100e0ae60f.png)

externallyManagedInitMethods value is 'init', initMethod value is 'init',  invokeCustomInitMethod method will not be invoked

I found that the value of the location setting is as follows:
![image](https://user-images.githubusercontent.com/22514893/154884645-0002f159-59b6-495a-9051-369f76c3b728.png)

Is there a problem here?

**Comment by MrocCyen** [comment]
![image](https://user-images.githubusercontent.com/22514893/154908064-c5144926-3189-4c96-bdbf-197c44dd327d.png)

Change it to this to get the right answer, is that right?

**Comment by snicoll** [comment]
@MrocCyen thanks but please do not share screenshots as it bloats the issue report. I've hidden those for now.

**Comment by MrocCyen** [comment]
> @MrocCyen thanks but please do not share screenshots as it bloats the issue report. I've hidden those for now.

Sorry, I'll pay attention next time. 

Excuse me, is it correct for me to fix it like that above?

**Comment by snicoll** [comment]
> Excuse me, is it correct for me to fix it like that above?

I haven't investigated this in detail so I can't say for sure.

**Comment by sbrannen** [comment]
I reproduced the described behavior with the following test class.

```java
@SpringJUnitConfig
class InitMethodTests {

	@Test
	void test() {
		SoftAssertions.assertSoftly(softly -> {
			softly.assertThat(PublicInitMethodBean.counter).as("public init-method").hasValue(1);
			softly.assertThat(PackageInitMethodBean.counter).as("package init-method").hasValue(1);
			softly.assertThat(PrivateInitMethodBean.counter).as("private init-method").hasValue(1);
		});
	}

	@Configuration
	static class Config {

		@Bean(initMethod = "init")
		Object publicInitMethodBean() {
			return new PublicInitMethodBean();
		}

		@Bean(initMethod = "init")
		Object packageInitMethodBean() {
			return new PackageInitMethodBean();
		}

		@Bean(initMethod = "init")
		Object privateInitMethodBean() {
			return new PrivateInitMethodBean();
		}
	}

	static class PublicInitMethodBean {
		static final AtomicInteger counter = new AtomicInteger();

		@PostConstruct
		public void init() {
			counter.incrementAndGet();
		}
	}

	static class PackageInitMethodBean {
		static final AtomicInteger counter = new AtomicInteger();

		@PostConstruct
		void init() {
			counter.incrementAndGet();
		}
	}

	static class PrivateInitMethodBean {
		static final AtomicInteger counter = new AtomicInteger();

		@PostConstruct
		private void init() {
			counter.incrementAndGet();
		}
	}

}
```

The "private init-method" assertion fails due to 2 invocations.

**Comment by sbrannen** [comment]
Superseded by #28113

**Comment by sbrannen** [comment]
### Related Issues

- #10762
- #27449

**Comment by sbrannen** [comment]
> Excuse me, is it correct for me to fix it like that above?

No, that would constitute a breaking change in behavior for `private` init/destroy methods with the same name declared at multiple levels in a class hierarchy.

We have decided to approach the fix differently, and I have therefore assigned this issue to myself.
